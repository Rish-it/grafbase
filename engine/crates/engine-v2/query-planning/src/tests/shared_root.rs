use crate::{
    tests::{read_schema, strdiff, TestOperation},
    OperationGraph,
};

const SCHEMA: &str = r###"
enum join__Graph {
  CATEGORY @join__graph(name: "category", url: "http://localhost:4200/shared-root/category")
  NAME @join__graph(name: "name", url: "http://localhost:4200/shared-root/name")
  PRICE @join__graph(name: "price", url: "http://localhost:4200/shared-root/price")
}

type Product
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
  category: String @join__field(graph: CATEGORY)
  name: String @join__field(graph: NAME)
  price: Float @join__field(graph: PRICE)
}

type Query
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  product: Product
  products: [Product]
}
"###;

#[test]
fn all_fields() {
    let schema = read_schema(SCHEMA);
    let mut operation = TestOperation::bind(
        &schema,
        r#"
        query {
          products {
            price
            category
            id
            name
          }
        }
    "#,
    );

    let mut graph = OperationGraph::new(&schema, &mut operation);
    insta::assert_snapshot!("all_fields-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());

    let before = graph.to_dot_graph();

    graph.prune_resolvers_not_leading_to_any_scalar_node();
    insta::assert_snapshot!(strdiff(&before, &graph.to_dot_graph()), @"");
}

#[test]
fn single_field() {
    let schema = read_schema(SCHEMA);
    let mut operation = TestOperation::bind(
        &schema,
        r#"
        query {
          products {
            price
          }
        }
    "#,
    );

    let mut graph = OperationGraph::new(&schema, &mut operation);
    insta::assert_snapshot!("single_field-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());

    let before = graph.to_dot_graph();

    graph.prune_resolvers_not_leading_to_any_scalar_node();
    insta::assert_snapshot!(strdiff(&before, &graph.to_dot_graph()), @r##"
    -    3 [ Root#category]
    -    4 [ products@Root#category]
    -    5 [ Root#name]
    -    6 [ products@Root#name]
    -    0 -> 3 [ label = "CreateChildResolver(1)" ]
    -    0 -> 3 [ label = "HasChildResolver" ]
    -    3 -> 4 [ label = "CanProvide(0)" ]
    -    4 -> 2 [ label = "Provides" ]
    -    0 -> 5 [ label = "CreateChildResolver(1)" ]
    -    0 -> 5 [ label = "HasChildResolver" ]
    -    5 -> 6 [ label = "CanProvide(0)" ]
    -    6 -> 2 [ label = "Provides" ]
    "##);
}
