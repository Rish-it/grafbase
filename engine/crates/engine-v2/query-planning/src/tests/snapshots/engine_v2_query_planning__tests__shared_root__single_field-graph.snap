---
source: engine/crates/engine-v2/query-planning/src/tests/shared_root.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"price\" ]\n    2 [ label = \"products\" ]\n    3 [ label = \"Root#category\", shape=box, color=blue ]\n    4 [ label = \"products@Root#category\", shape=box, style=dashed, color=blue ]\n    5 [ label = \"Root#name\", shape=box, color=blue ]\n    6 [ label = \"products@Root#name\", shape=box, style=dashed, color=blue ]\n    7 [ label = \"Root#price\", shape=box, color=blue ]\n    8 [ label = \"products@Root#price\", shape=box, style=dashed, color=blue ]\n    9 [ label = \"price@Root#price\", shape=box, style=dashed, color=blue ]\n    0 -> 2 [ label = \"\" ]\n    0 -> 3 [ label = <<b>1</b>>, color=blue,fontcolor=blue ]\n    0 -> 3 [ label = \"\", style=dashed ]\n    3 -> 4 [ label = \"\", color=blue,fontcolor=blue ]\n    4 -> 2 [ label = \"\", color=turquoise ]\n    0 -> 5 [ label = <<b>1</b>>, color=blue,fontcolor=blue ]\n    0 -> 5 [ label = \"\", style=dashed ]\n    5 -> 6 [ label = \"\", color=blue,fontcolor=blue ]\n    6 -> 2 [ label = \"\", color=turquoise ]\n    0 -> 7 [ label = <<b>1</b>>, color=blue,fontcolor=blue ]\n    0 -> 7 [ label = \"\", style=dashed ]\n    7 -> 8 [ label = \"\", color=blue,fontcolor=blue ]\n    8 -> 2 [ label = \"\", color=turquoise ]\n    2 -> 1 [ label = \"\" ]\n    8 -> 9 [ label = \"\", color=blue,fontcolor=blue ]\n    9 -> 1 [ label = \"\", color=turquoise ]\n}\n"
---
digraph {
    0 [ root]
    1 [ price]
    2 [ products]
    3 [ Root#category]
    4 [ products@Root#category]
    5 [ Root#name]
    6 [ products@Root#name]
    7 [ Root#price]
    8 [ products@Root#price]
    9 [ price@Root#price]
    0 -> 2 [ label = "Field" ]
    0 -> 3 [ label = "CreateChildResolver(1)" ]
    0 -> 3 [ label = "HasChildResolver" ]
    3 -> 4 [ label = "CanProvide(0)" ]
    4 -> 2 [ label = "Provides" ]
    0 -> 5 [ label = "CreateChildResolver(1)" ]
    0 -> 5 [ label = "HasChildResolver" ]
    5 -> 6 [ label = "CanProvide(0)" ]
    6 -> 2 [ label = "Provides" ]
    0 -> 7 [ label = "CreateChildResolver(1)" ]
    0 -> 7 [ label = "HasChildResolver" ]
    7 -> 8 [ label = "CanProvide(0)" ]
    8 -> 2 [ label = "Provides" ]
    2 -> 1 [ label = "Field" ]
    8 -> 9 [ label = "CanProvide(0)" ]
    9 -> 1 [ label = "Provides" ]
}
